File: java/src/test/java/io/github/moctave/weftspace/TestBuilder.java
// Copyright (c) 2026 by mOctave
//
// This program is free software: you can redistribute it and/or modify it under the
// terms of the GNU Affero General Public License as published by the Free Software
// Foundation, either version 3 of the License, or (at your option) any later version.
//
// This program is distributed in the hope that it will be useful, but WITHOUT ANY
// WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A
// PARTICULAR PURPOSE. See the GNU Affero General Public License for more details.
//
// You should have received a copy of the GNU Affero General Public License along with
// this program. If not, see <https://www.gnu.org/licenses/>.

package io.github.moctave.weftspace;

import static org.junit.jupiter.api.Assertions.*;

import java.util.List;

import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;

public class TestBuilder {
	private DataNode testNoArgs;
	private DataNode testStringArgs;
	private DataNode testIntegerArgs;
	private DataNode testDoubleArgs;

	@BeforeEach
	public void init() {
		testDoubleArgs = new DataNode("doubargs", null, List.of("0.", "-0.8", "15.36", "3.14159"), List.of());
		testStringArgs = new DataNode("strargs", null, List.of("foo", "bar"), List.of());
		testIntegerArgs = new DataNode("intargs", null, List.of("1", "2", "3", "-1"), List.of(testDoubleArgs));
		testNoArgs = new DataNode("argless", null, List.of(), List.of(testStringArgs, testIntegerArgs));
	}


	// MARK: buildString()
	@Test
	public void testBuildStringNoArgs() {
		try {
			Builder.buildString(testNoArgs, 0);
			fail();
		} catch (BuilderException e) {
			// Expected result
		}
	}

	@Test
	public void testBuildStringInRange() {
		try {
			assertEquals("foo", Builder.buildString(testStringArgs, 0));
			assertEquals("3.14159", Builder.buildString(testDoubleArgs, 3));
		} catch (BuilderException e) {
			fail();
		}
	}

	@Test
	public void testBuildStringOutOfRange() {
		try {
			Builder.buildString(testStringArgs, 2);
			fail();
		} catch (BuilderException e) {
			// Expected result
		}
	}



	// MARK: buildInt()
	@Test
	public void testBuildIntNoArgs() {
		try {
			Builder.buildInt(testNoArgs, 0);
			fail();
		} catch (BuilderException e) {
			// Expected result
		}
	}

	@Test
	public void testBuildIntInRange() {
		try {
			assertEquals(3, Builder.buildInt(testIntegerArgs, 2));
			assertEquals(-1, Builder.buildInt(testIntegerArgs, 3));
		} catch (BuilderException e) {
			fail();
		}
	}

	@Test
	public void testBuildIntOutOfRange() {
		try {
			Builder.buildInt(testIntegerArgs, 4);
			fail();
		} catch (BuilderException e) {
			// Expected result
		}
	}

	@Test
	public void testBuildIntStringValue() {
		try {
			Builder.buildInt(testStringArgs, 0);
			fail();
		} catch (BuilderException e) {
			// Expected result
		}
	}

	@Test
	public void testBuildIntDoubleValue() {
		try {
			Builder.buildInt(testDoubleArgs, 0);
			fail();
		} catch (BuilderException e) {
			// Expected result
		}
	}




	// MARK: buildDouble()
	@Test
	public void testBuildDoubleNoArgs() {
		try {
			Builder.buildDouble(testNoArgs, 0);
			fail();
		} catch (BuilderException e) {
			// Expected result
		}
	}

	@Test
	public void testBuildDoubleInRange() {
		try {
			assertEquals(0., Builder.buildDouble(testDoubleArgs, 0));
			assertEquals(3.14159, Builder.buildDouble(testDoubleArgs, 3), 1e-6);
		} catch (BuilderException e) {
			fail();
		}
	}

	@Test
	public void testBuildDoubleOutOfRange() {
		try {
			Builder.buildDouble(testDoubleArgs, 4);
			fail();
		} catch (BuilderException e) {
			// Expected result
		}
	}

	@Test
	public void testBuildDoubleStringValue() {
		try {
			Builder.buildDouble(testStringArgs, 0);
			fail();
		} catch (BuilderException e) {
			// Expected result
		}
	}

	@Test
	public void testBuildDoubleIntValue() {
		try {
			assertEquals(1., Builder.buildDouble(testIntegerArgs, 0));
		} catch (BuilderException e) {
			fail();
		}
	}



	// MARK: buildLong()
	@Test
	public void testBuildLongNoArgs() {
		try {
			Builder.buildLong(testNoArgs, 0);
			fail();
		} catch (BuilderException e) {
			// Expected result
		}
	}

	@Test
	public void testBuildLongInRange() {
		try {
			assertEquals(3l, Builder.buildLong(testIntegerArgs, 2));
			assertEquals(-1l, Builder.buildLong(testIntegerArgs, 3));
		} catch (BuilderException e) {
			fail();
		}
	}

	@Test
	public void testBuildLongOutOfRange() {
		try {
			Builder.buildLong(testIntegerArgs, 4);
			fail();
		} catch (BuilderException e) {
			// Expected result
		}
	}

	@Test
	public void testBuildLongStringValue() {
		try {
			Builder.buildLong(testStringArgs, 0);
			fail();
		} catch (BuilderException e) {
			// Expected result
		}
	}

	@Test
	public void testBuildLongDoubleValue() {
		try {
			Builder.buildLong(testDoubleArgs, 0);
			fail();
		} catch (BuilderException e) {
			// Expected result
		}
	}



	// MARK: search()
	@Test
	public void testSearchNoMatch() {
		assertNull(Builder.search(
			new DataNode("a", null, List.of("q", "r", "s"), List.of()),
			testNoArgs)
		);
	}

	@Test
	public void testSearchExactMatch() {
		assertEquals(testStringArgs, Builder.search(testStringArgs, testNoArgs));
	}

	@Test
	public void testSearchInexactMatch() {
		assertEquals(testIntegerArgs, Builder.search(
			new DataNode("intargs", null, List.of("1", "0", "not even an int"), List.of()),
			testNoArgs)
		);
	}


	@Test
	public void testSearchCloseFailure() {
		assertNull(Builder.search(
			new DataNode("intargs", null, List.of(), List.of()),
			testNoArgs)
		);

		assertNull(Builder.search(
			new DataNode("intargs", null, List.of("."), List.of()),
			testNoArgs)
		);

		assertNull(Builder.search(
			new DataNode("not int args", null, List.of("1"), List.of()),
			testNoArgs)
		);
	}

	@Test
	public void testSearchSecondLevelMatch() {
		assertNull(Builder.search(testDoubleArgs, testNoArgs));
	}
}

File: python/tests/test_builder.py
# Copyright (c) 2026 by mOctave
#
# This program is free software: you can redistribute it and/or modify it under the
# terms of the GNU Affero General Public License as published by the Free Software
# Foundation, either version 3 of the License, or (at your option) any later version.
#
# This program is distributed in the hope that it will be useful, but WITHOUT ANY
# WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A
# PARTICULAR PURPOSE. See the GNU Affero General Public License for more details.
#
# You should have received a copy of the GNU Affero General Public License along with
# this program. If not, see <https://www.gnu.org/licenses/>.

import unittest

from weftspace import Builder, BuilderError, DataNode

class TestBuilder(unittest.TestCase):
	"""Tests for the various methods of Builder"""

	no_args: DataNode
	string_args: DataNode
	int_args: DataNode
	float_args: DataNode

	def setUp(self) -> None:
		print("PRINT ME")
		self.float_args = DataNode("doubargs", None, ["0.", "-0.8", "15.36", "3.14159"], [])
		self.string_args = DataNode("strargs", None, ["foo", "bar"], [])
		self.int_args = DataNode("intargs", None, ["1", "2", "3", "-1"], [self.float_args])
		self.no_args = DataNode("argless", None, [], [self.string_args, self.int_args])
		return super().setUp()

	def test_pass(self):
		return

	# MARK: build_string()
	def test_build_string_no_args(self) -> None:
		try:
			Builder.build_string(self.no_args, 0)
			self.fail()
		except BuilderError:
			# Expected Result
			pass
	
	def test_build_string_in_range(self) -> None:
		try:
			self.assertEqual("foo", Builder.build_string(self.string_args, 0))
			self.assertEqual("3.14159", Builder.build_string(self.float_args, 3))
		except BuilderError:
			self.fail()

	def test_build_string_out_of_range(self) -> None:
		try:
			Builder.build_string(self.string_args, 2)
			self.fail()
		except BuilderError:
			# Expected Result
			pass



	# MARK: build_int()
	def test_build_int_no_args(self) -> None:
		try:
			Builder.build_int(self.no_args, 0)
			self.fail()
		except BuilderError:
			# Expected Result
			pass
	
	def test_build_int_in_range(self) -> None:
		try:
			self.assertEqual(3, Builder.build_int(self.int_args, 2))
			self.assertEqual(-1, Builder.build_int(self.int_args, 3))
		except BuilderError:
			self.fail()

	def test_build_int_out_of_range(self) -> None:
		try:
			Builder.build_int(self.int_args, 4)
			self.fail()
		except BuilderError:
			# Expected Result
			pass

	def test_build_int_string_value(self) -> None:
		try:
			Builder.build_int(self.string_args, 0)
			self.fail()
		except BuilderError:
			# Expected Result
			pass

	def test_build_int_float_value(self) -> None:
		try:
			Builder.build_int(self.float_args, 0)
			self.fail()
		except BuilderError:
			# Expected Result
			pass



	# MARK: build_float()
	def test_build_float_no_args(self) -> None:
		try:
			Builder.build_float(self.no_args, 0)
			self.fail()
		except BuilderError:
			# Expected Result
			pass
	
	def test_build_float_in_range(self) -> None:
		try:
			self.assertAlmostEqual(0., Builder.build_float(self.float_args, 0))
			self.assertAlmostEqual(3.14159, Builder.build_float(self.float_args, 3))
		except BuilderError:
			self.fail()

	def test_build_float_out_of_range(self) -> None:
		try:
			Builder.build_float(self.float_args, 4)
			self.fail()
		except BuilderError:
			# Expected Result
			pass

	def test_build_float_string_value(self) -> None:
		try:
			Builder.build_float(self.string_args, 0)
			self.fail()
		except BuilderError:
			# Expected Result
			pass

	def test_build_float_int_value(self) -> None:
		try:
			self.assertAlmostEqual(1., Builder.build_float(self.int_args, 0))
		except BuilderError:
			self.fail()
			pass



	# search()
	def test_search_no_match(self) -> None:
		self.assertEqual(None, Builder.search(DataNode("a", None, ["q", "r", "s"], []), self.no_args))


	def test_search_exact_match(self) -> None:
		self.assertEqual(self.string_args, Builder.search(self.string_args, self.no_args))

	def test_search_inexact_match(self) -> None:
		self.assertEqual(self.int_args,
			Builder.search(DataNode("intargs", None, ["1", "0", "not even an int"], []), self.no_args))


	def test_search_close_failure(self) -> None:
		self.assertEqual(None,
			Builder.search(DataNode("intargs", None, [], []), self.no_args))
		self.assertEqual(None,
			Builder.search(DataNode("intargs", None, ["."], []), self.no_args))
		self.assertEqual(None,
			Builder.search(DataNode("not int args", None, ["1"], []), self.no_args))

	def test_search_second_level_match(self) -> None:
		self.assertEqual(None, Builder.search(self.float_args, self.no_args))



if __name__ == "__main__":
	unittest.main()
